assembly/src/ast/module.rs

```diff
@@ -1,322 +1,723 @@
+/// Construction
+impl Module {
+    /// Creates a new [Module] with the specified `kind` and fully-qualified path, e.g.
+    /// `std::math::u64`.
+    pub fn new(kind: ModuleKind, path: LibraryPath) -> Self {
+        Self {
+            span: Default::default(),
+            source_file: None,
+            docs: None,
+            path,
+            kind,
+            imports: Default::default(),
+            procedures: Default::default(),
+        }
+    }
+
+    /// An alias for creating the default, but empty, `#kernel` [Module].
+    pub fn new_kernel() -> Self {
+        Self::new(ModuleKind::Kernel, LibraryNamespace::Kernel.into())
+    }
+
+    /// An alias for creating the default, but empty, `#exec` [Module].
+    pub fn new_executable() -> Self {
+        Self::new(ModuleKind::Executable, LibraryNamespace::Exec.into())
+    }
+
+    /// Builds this [Module] with the given source file in which it was defined.
     ///
-    /// A module consists of internal and exported procedures but does not contain a body.
-    pub fn new(
-        local_procs: Vec<ProcedureAst>,
-        reexported_procs: Vec<ProcReExport>,
-        docs: Option<String>,
-    ) -> Result<Self, ParsingError> {
-        if local_procs.len() > MAX_LOCAL_PROCS {
-            return Err(ParsingError::too_many_module_procs(local_procs.len(), MAX_LOCAL_PROCS));
+    /// When a source file is given, diagnostics will contain source code snippets.
+    pub fn with_source_file(mut self, source_file: Option<Arc<SourceFile>>) -> Self {
+        self.source_file = source_file;
+        self
+    }
+
+    /// Specifies the source span in the source file in which this module was defined, that covers
+    /// the full definition of this module.
+    pub fn with_span(mut self, span: SourceSpan) -> Self {
+        self.span = span;
+        self
+    }
+
+    /// Like [Module::with_source_file], but does not require ownership of the [Module].
+    pub fn set_source_file(&mut self, source_file: Arc<SourceFile>) {
+        self.source_file = Some(source_file);
+    }
+
+    /// Sets the [LibraryPath] for this module
+    pub fn set_path(&mut self, path: LibraryPath) {
+        self.path = path;
+    }
+
+    /// Sets the [LibraryNamespace] for this module
+    pub fn set_namespace(&mut self, ns: LibraryNamespace) {
+        self.path.set_namespace(ns);
+    }
+
+    /// Sets the documentation for this module
+    pub fn set_docs(&mut self, docs: Option<Span<String>>) {
+        self.docs = docs;
+    }
+
+    /// Like [Module::with_span], but does not require ownership of the [Module].
+    pub fn set_span(&mut self, span: SourceSpan) {
+        self.span = span;
+    }
+
+    /// Defines a procedure, raising an error if the procedure is invalid, or conflicts with a
+    /// previous definition
+    pub fn define_procedure(&mut self, export: Export) -> Result<(), SemanticAnalysisError> {
+        if self.is_kernel() && matches!(export, Export::Alias(_)) {
```

Comment:
I'm starting a new thread to reply to your comment here, because the previous one got lost due to a force push when I rebased AFAICT.

> The way I think about it is that any procedure exported from the kernel module is syscall-able - this includes both procedures defined in the module itself as well as re-exported from other modules. I'm not sure the above restriction is actually meaningful because instead of:
>
> NOTE: Edits made by pauls below to correct some syntax errors, see [here](https://github.com/0xPolygonMiden/miden-vm/pull/1277#discussion_r1546148844) for the original comment, though GitHub may not you see it anymore
>
> ```
> # baz.masm
> use.foo
> 
> export foo::bar
>```
>
> We can always write:
>
> ```
> # baz.masm
> use.foo
>    
> export.bar
>     exec.foo::bar
> end
> ```

From the perspective of the user looking at what is exported from a kernel module, yes, but there is an important distinction here: `foo` is _not_ a kernel module, it cannot contain any elements that are restricted outside of kernel modules. Furthermore, we cannot guarantee that `foo::bar` in this case was written in such a way that it adheres to the ABI (implicit or otherwise) expected of syscalls.

In short, the latter of your two code samples is actually what we _should_ require be done in this case, as the author of the kernel module is explicitly signing off on `foo::bar` being valid as a syscall. If we ultimately end up adding support for `#[inline]` attributes, then there is zero difference in terms of how the two are compiled to MAST in the end.

I think fundamentally, it does not make sense to allow defining syscalls in libraries (and that is implicitly what is being done if we state that re-exporting a library procedure from a kernel makes it a syscall). That would be completely non-obvious when looking at the library module in question (only when reviewing the kernel would it become clear that it can be used that way). I think we should require that all syscalls be defined _in_ the kernel module, even if that feels redundant in some cases (by needing to define wrappers as you demonstrated).

That said, if we move to a model where all procedures declare _how_ they can be invoked (e.g. syscall, call, or exec - essentially the most fundamental aspect of the procedure ABI), then we could loosen this restriction (instead, we would require that any procedures re-exported from the kernel be of `syscall` type). For now though, we just simply don't have enough information for this to be a good idea IMO.

Reply:
> I think fundamentally, it does not make sense to allow defining syscalls in libraries

I think I probably agree with this. But just to note, this is not how things worked prior to this refactoring - i.e., it was perfectly fine to use `caller` instruction in any module and if it was used incorrectly, this would be caught at the assembly time (i.e., when we actually built MAST for the kernel module).

Not an issue for the transaction kernel we have now as we don't use `caller` in any of the dependencies.

So, basically - I'm fine with keeping this restriction (i.e., not allowing re-exporting procedures from the kernel module).

> That said, if we move to a model where all procedures declare _how_ they can be invoked (e.g. syscall, call, or exec - essentially the most fundamental aspect of the procedure ABI), then we could loosen this restriction (instead, we would require that any procedures re-exported from the kernel be of `syscall` type). For now though, we just simply don't have enough information for this to be a good idea IMO.

Agreed - this is probably the right long-term solution.

Reply:
> I think I probably agree with this. But just to note, this is not how things worked prior to this refactoring - i.e., it was perfectly fine to use caller instruction in any module and if it was used incorrectly, this would be caught at the assembly time (i.e., when we actually built MAST for the kernel module).

The main issue here is that prior this refactoring, it was assumed that all code was being compiled together - that won't be the case moving forward. Procedures may be referenced that we don't have sources for, only MAST. In such cases, we won't have the opportunity to validate things that way.

That said, there are two caveats I want to mention:

1. This PR doesn't fully implement support for linking in raw MAST. Instead we assume that anything referenced by MAST root is provided to the VM out-of-band, and that anything referenced by name will be provided in source form - but that is purely due to MAST currently lacking metadata we need to be able to reference procedures in the MAST by name. Ostensibly this would be provided by the package format (or as metadata associated with a MAST artifact, either is fine). However, support for this is baked in to how the assembler is structured (in particular the way the module graph and procedure cache are designed), so once we have that metadata, making it possible to assemble a mixture of MASM and MAST will be trivial to implement as an extension of the `Assembler` API.
2. By representing the capability set of a procedure as metadata paired with a given MAST root, we could support validating that all procedures exported by (or called from) the kernel module, are valid in that context.

That said, if we need to support the old behavior in the interim, we certainly can - but it sounds like we probably don't need to (which would be great for us!), but you'll have to confirm.